sound/soc/sof/intel/hda-loader.c

```diff
@@ -373,6 +376,142 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
+
+		bdl[2] = cpu_to_le32(size);
+
+		size -= size;
+		bdl[3] = (size || !with_ioc) ? 0 : cpu_to_le32(0x01);
+
+		bdl += 4;
+		frags++;
+	}
+
+	return frags;
+}
+
+static void cl_skl_cldma_stream_run(struct snd_sof_dev *sdev, bool enable)
+{
+	unsigned char val;
+	int timeout;
+
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_RUN_MASK, CL_SD_CTL_RUN(enable));
+
+	udelay(3);
+	timeout = 300;
+	do {
+		/* waiting for hardware to report the stream Run bit set */
+		val = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR,
+				       SKL_ADSP_REG_CL_SD_CTL) &
+				       CL_SD_CTL_RUN_MASK;
+		if (enable && val)
+			break;
+		else if (!enable && !val)
+			break;
+		udelay(3);
+	} while (--timeout);
+
+	if (timeout == 0)
+		dev_err(sdev->dev, "Failed to set Run bit=%d enable=%d\n",
+			val, enable);
+}
+
+/*
+ * Setup controller
+ * Configure the registers to update the dma buffer address and
+ * enable interrupts.
+ * Note: Using the channel 1 for transfer
+ */
+static void cl_skl_cldma_setup_controller(struct snd_sof_dev *sdev,
+					  struct snd_dma_buffer *dmab_bdl,
+					  unsigned int max_size, u32 count)
+{
+	/* make sure Run bit is cleared before setting stream register */
+	cl_skl_cldma_stream_run(sdev, 0);
+
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_IOCE_MASK, CL_SD_CTL_IOCE(0));
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_FEIE_MASK, CL_SD_CTL_FEIE(0));
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_DEIE_MASK, CL_SD_CTL_DEIE(0));
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_STRM_MASK, CL_SD_CTL_STRM(0));
+
+	snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_BDLPL,
+			  CL_SD_BDLPLBA(0));
+	snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_BDLPU, 0);
+
+	snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CBL, 0);
+	snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_LVI, 0);
+
+	/* setting the stream register */
+	snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_BDLPL,
+			  CL_SD_BDLPLBA(dmab_bdl->addr));
+	snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_BDLPU,
+			  CL_SD_BDLPUBA(dmab_bdl->addr));
+
+	snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CBL,
+			  max_size);
+	snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_LVI,
+			  count - 1);
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_IOCE_MASK, CL_SD_CTL_IOCE(1));
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_FEIE_MASK, CL_SD_CTL_FEIE(1));
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_DEIE_MASK, CL_SD_CTL_DEIE(1));
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_STRM_MASK,
+				CL_SD_CTL_STRM(FW_CL_STREAM_NUMBER));
```

Comment:
Last 4 write are all for same register. They should be combined.

Reply:
Same as up.